File: 03_04_QueuesUsingStacks.java
class MyQueue {
    
    Deque<Integer> stack1 = new ArrayDeque<>();
    Deque<Integer> stack2 = new ArrayDeque<>();

    public MyQueue() {
    }
    
    public void push(int x) {
        stack1.push(x);
    }
    
    public int pop() {
        moveIfEmpty();
        
        return stack2.pop();
    }
    
    public int peek() {
        moveIfEmpty();
        
        return stack2.peek();
    }
    
    public boolean empty() {
        return stack1.isEmpty() && stack2.isEmpty();
    }
    
    private void moveIfEmpty() {
        if (stack2.isEmpty()) {
            while(!stack1.isEmpty()) {
                stack2.push(stack1.pop());
            }
        }
    }
}

/**
 * Your MyQueue object will be instantiated and called as such:
 * MyQueue obj = new MyQueue();
 * obj.push(x);
 * int param_2 = obj.pop();
 * int param_3 = obj.peek();
 * boolean param_4 = obj.empty();
 */

File: 03__04_QueuesUsingStacks.py
class MyQueue:

    def __init__(self):
        self.stack1 = []
        self.stack2 = []
        self.top = 0
        

    def push(self, x: int) -> None:
        if not self.stack1:
            self.top = x
        self.stack1.append(x)
        print("stack1",self.stack1)
        
        

    def pop(self) -> int:
        if not self.stack2:
            while self.stack1:
                    self.stack2.append(self.stack1.pop())
        print("stack2",self.stack2)
        return self.stack2.pop()
        

    def peek(self) -> int:
        if self.stack2:
            return self.stack2[-1]
        else: return self.top
        

    def empty(self) -> bool:
        return len(self.stack2)==0 and len(self.stack1)==0
